text/006-local-js-dependencies.md

```diff
@@ -0,0 +1,347 @@
+  possible for authors to manually run the TypeScript compiler by hand for these
+  use cases, though.
+
+* The relatively popular `--no-modules` flag is proposed to be deprecated in
+  favor of a `--browser` flag, which itself will have a breaking change relative
+  to today. It's thought though that `--browser` is only very rarely used so is
+  safe to break, and it's also thought that we'll want to avoid breaking
+  `--no-modules` as-is today.
+
+* JS files are imported relative to the crate root rather than the file doing
+  the importing, unlike `mod` statements in Rust. It's thought that we can't get
+  file-relative imports working with stable `proc_macro` APIs today, but if the
+  implementation can manage to finesse it this RFC would propose instead making
+  file imports relative to the current file instead of crate root.
+
+* Local JS snippets are required to be written in ES module syntax. This may be
+  a somewhat opinionated stance, but it's intended to make it easier to add
+  future features to `wasm-bindgen` while continuing to work with JS.
+
+# Rationale and Alternatives
+[alternatives]: #rationale-and-alternatives
+
+Currently there aren't any competing designs solving this problem, so there
+aren't many known major alternatives. There's a number of alternatives to
+various smaller decisions in this RFC, but it's hoped that the various
+alternatives are listed or discussed inlined.
+
+The major rationale for this RFC is empowering this use case of "seamless local
+JS snippets" **at all**. The RFC proposes to start out with only including files
+as opposed to small JS snippets themselves (like `stdweb`'s own `js!` macro)
+because files are structured as ES modules which is what `#[wasm_bindgen]`
+already supports well and will also eventually have support natively in
+WebAssembly itself. While `#[wasm_bindgen]` may one day have a `js!`-like macro
+built-in, it's hoped that we can start out with a purely library-based solution
+like `stdweb`, iterate on it, and consider this question later.
+
+One alternative for ES modules is to simply concatenate all JS together. This
+way we wouldn't have to parse anything but we'd instead just throw everything
+into one file. The downside of this approach, however, is that it can easily
+lead to namespacing conflicts and it also forces everyone to agree on module
+formats and runs the risk of forcing the module format of the final product.
+
+Another alternative to emitting small files at wasm-bindgen time is to instead
+unpack all files at *runtime* by leaving them in custom sections of the wasm
+executable. This in turn, however, may violate some CSP settings (particularly
+strict ones).
+
+# Unresolved Questions
+[unresolved]: #unresolved-questions
+
+- Is it necessary to support `--nodejs` initially?
+
+- Is it necessary to support local JS imports in local JS snippets initially?
+
+- Are there known parsers of JS ES modules today? Are we forced to include a
```

Comment:
If we simply want to support "JS files importing other JS files", I think a very minimal parser should be fine.

If we want to support ES6 module concatenation, then we probably need a full parser. But as I said above, I think we should try to use existing solutions rather than making our own (unless there's a good reason to make our own).

Reply:
One aspect I'm worried about is that we have to also parse global declarations like `let x = 3;` which I fear will pull in a full JS parser to handle that.

Reply:
If we want to support concatenation built-in to wasm-bindgen, then yeah we need a full parser. But if it's just parsing `import` stuff to find other `.js` files to import, that's a different story.

Actually, come to think of it, how will incremental building work with `.js` files importing other `.js` files? If a user edits a `.js` file and adds a new `import`, will that change be picked up?

Reply:
My hope is to use `include_str!` or similar to inject a dependency via rustc on the actual file itself, so Cargo will rerun the build of a crate if the `.js` file is edited. For `import` it sort of naturally falls out from the rest of this RFC. It's either supported by parsing the file (so we'll re-parse and figure it out), supported by pulling in everything (but has the drawback that we can't handle newly added files), or not supported right now (which is what this RFC currently proposes)

Reply:
AFAIK we don't actually have much reason to worry about `package.json` in `wasm-bindgen`, only `wasm-pack`. In `wasm-pack` everything is unconditionally done all the time (as it's so fast), but if we were to need to parse things in `wasm-bindgen` this'd be how we do it!

Reply:
So how is that intended to work (especially with transitive crates)? `wasm-pack` can't access transitive crates, right?

I know that's going a bit beyond this RFC, but since it relates to dynamic `import()` I think it's somewhat relevant.

Reply:
Hm I may be missing some context to the question, can you be a bit more specific about what you're curious will work?

To try to answer ahead of time, though, you're right that `wasm-pack` (and the `wasm-bindgen` CLI) don't have access to the particulars of the crate graph. The `#[wasm_bindgen]` macro does, however, have access! To that end the `#[wasm_bindgen]` macro can encode information into the custom section (which gets concatenated) about how to find JS files. It can either serialize them directly into the custom section or leave filesystem paths in the custom section which immediately get removed during `wasm-bindgen`-the-CLI. 

In that sense the outer tools don't have direct access to transitive crates, but the macro can encode and provide all the information they should need to operate.

Reply:
> To try to answer ahead of time, though, you're right that wasm-pack (and the wasm-bindgen CLI) don't have access to the particulars of the crate graph.

That's what I'm concerned about. In order to make things work correctly, it is necessary for `package.json` to work with transitive crates. It's not that important for this RFC, but it's important in general.

So if `wasm-bindgen` doesn't include the `package.json` in the custom section, how will `wasm-pack` be able to handle `package.json` in transitive crates?

Maybe we should move this conversation to another issue, since it seems to be getting quite off-topic.

Reply:
Oh sorry my point is that I agree that transitive crates should always work, and the `#[wasm_bindgen]` macro is how we'll solve that. If we need to ship `package.json` (or anything else) from dependencies to the final artifact, we'll do it in the `#[wasm_bindgen]` macro

(moving to a separate issue is also fine by me